### PR TITLE
Remove default timeout from PGNCCL::Options ctor

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112611
* __->__ #112555

Providing this timeout to the Options ctor is overriding user-provided
values in cases where is_high_priority_stream is set.

### CODE REVIEW

**Review by fduwjj — APPROVED** [review]

**Review by H-Huang — COMMENTED** [review]
I think we will need to move this out of `else` block for the user set timeout to get respected

https://github.com/pytorch/pytorch/blob/84f59d893a029ebb804ba344e48fb2c2c56a213d/torch/distributed/distributed_c10d.py#L1291

The issue is that the ProcessGroupOptions timeout will be overridden if the user happens to set it...

**Comment by wconstab** [comment]
@H-Huang any issue with landing this as-is first, and iterating the next PR to handle the duplicate timeout args?  (also, pls comment there i'm wondering what the best thing to do is)

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -1898,7 +1898,7 @@ void ProcessGroupNCCL::workEnqueue(
 }
 
 ProcessGroupNCCL::Options::Options(bool is_high_priority_stream)
-    : Backend::Options(NCCL_BACKEND_NAME, kProcessGroupNCCLDefaultTimeout),
+    : Backend::Options(NCCL_BACKEND_NAME),
```
Comment (H-Huang) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1901]: Okay I think I understand the confusion now, we are basically never using this pybind constructor (maybe it is used in some tests where we instantiate ProcessGroupNCCL directly rather than through `init_pg`):

https://github.com/pytorch/pytorch/blob/a1ab22b81d74a26873243d766d78724b3b70d570/torch/csrc/distributed/c10d/init.cpp#L2246-L2255

Instead in `init_pg` we always pass in `options` into the constructor so we always go through this path:

https://github.com/pytorch/pytorch/blob/a1ab22b81d74a26873243d766d78724b3b70d570/torch/csrc/distributed/c10d/init.cpp#L2239-L2245

So basically this change will change back to the original behavior of 30mins timeout and wont use `kProcessGroupNCCLDefaultTimeout`. But we need this change and the change in the stack above for the desired behavior, so im stamping it

**Review by H-Huang — APPROVED** [review]

**Comment by wconstab** [comment]
Thanks @H-Huang for helping me understand what's going on here.  This PR is actually wrong.  It happened to help a particular user that was calling `dist.init_process_group` and specifying both a `timeout=30min` kwarg, and a `pg_options` value, where no timeout was specified for the pg_options.

In that case, the `timeout` kwarg was being totally ignored, due to the bug that is fixed in #112611, and the default timeout on the options was taking precedence.  

The change in this PR means that for anyone constructing a PGNccl::Options and feeding that to PGNCCL's ctor, they would get the 'Backend::Options' timeout value (still 30 min) rather than the intended 'PGNccl::Options' timeout value (supposed to be 10 min).
